xmtp_mls/src/groups/commit_log.rs

```diff
@@ -344,46 +316,248 @@ where
+        entry.group_id != validation_info.requested_group_id
+            || entry.commit_sequence_id == 0
+            || entry.commit_sequence_id <= validation_info.latest_stored_commit_sequence_id
+            || (is_applied
+                && !validation_info
+                    .latest_applied_epoch_authenticator
+                    .is_empty()
+                && entry.last_epoch_authenticator
+                    != validation_info.latest_applied_epoch_authenticator)
+            || (is_applied
+                && entry.applied_epoch_number != validation_info.latest_applied_epoch_number + 1)
+    }
+
+    pub async fn check_forked_state(
+        &mut self,
+    ) -> Result<HashMap<Vec<u8>, ForkedStateCheckResult>, CommitLogError> {
+        let conn = &self.context.db();
+        let conversation_ids_for_forked_state_check =
+            conn.get_conversation_ids_for_remote_log_download()?;
+
+        let mut forked_state_check_results = HashMap::new();
+
+        for conversation_id in conversation_ids_for_forked_state_check {
+            let fork_result = self.check_conversation_fork_state(conn, &conversation_id)?;
+            forked_state_check_results.insert(conversation_id, fork_result);
+        }
+
+        Ok(forked_state_check_results)
+    }
+
+    fn check_conversation_fork_state(
+        &self,
+        conn: &impl DbQuery,
+        conversation_id: &[u8],
+    ) -> Result<ForkedStateCheckResult, CommitLogError> {
+        // Get cursors for this conversation
+        let fork_check_local_cursor = conn.get_last_cursor_for_id(
+            conversation_id,
+            xmtp_db::refresh_state::EntityKind::CommitLogForkCheckLocal,
+        )?;
+        let fork_check_remote_cursor = conn.get_last_cursor_for_id(
+            conversation_id,
+            xmtp_db::refresh_state::EntityKind::CommitLogForkCheckRemote,
+        )?;
+
+        // Get local and remote commit logs
+        let local_logs = conn.get_local_commit_log_after_cursor(
+            conversation_id,
+            fork_check_local_cursor,
+            LocalCommitLogOrder::DescendingByRowid,
+        )?;
+        let remote_logs =
+            conn.get_remote_commit_log_after_cursor(conversation_id, fork_check_remote_cursor)?;
+
+        // Check each local log against remote logs for matching commit_sequence_id
+        for local_log in &local_logs {
+            if let Some(matching_remote_log) =
+                self.find_matching_remote_log(&remote_logs, local_log.commit_sequence_id)
+            {
+                // Found a matching commit_sequence_id - check if forked
+                let is_forked = local_log.applied_epoch_authenticator
+                    != matching_remote_log.applied_epoch_authenticator;
+
+                if is_forked {
+                    println!("Forked state check result: {:?}", matching_remote_log);
+                    println!("Local log: {:?}", local_log);
```

Comment:
Should these be `tracing` calls rather than prints?

Reply:
updated 97664f22ff8903ec107254fd2ba1eac7db1e50a9